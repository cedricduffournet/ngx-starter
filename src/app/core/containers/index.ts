export * from './toaster';
export * from './http-progress-bar/http-progress-bar.component';
