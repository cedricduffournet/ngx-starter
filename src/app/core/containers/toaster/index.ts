export * from './toaster.component';
export * from './toaster.model';
