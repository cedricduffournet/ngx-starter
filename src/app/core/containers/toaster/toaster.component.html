<toaster-container [toasterconfig]="toasterConfig"></toaster-container>
