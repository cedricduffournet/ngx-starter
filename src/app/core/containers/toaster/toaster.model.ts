export const defaultConfig = {
  showCloseButton: false,
  timeout: 5000
};

export interface ToasterParams {
  type: string;
  title: string;
  body: string;
  clear?: boolean;
  timeout?: number | null;
}
