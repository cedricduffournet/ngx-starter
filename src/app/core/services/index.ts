export * from './http-progress.interceptor';
export * from './http.service';
