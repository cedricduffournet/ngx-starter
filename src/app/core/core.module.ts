import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { environment } from '@env/environment';
import {
  HttpClientModule,
  HttpClient,
  HTTP_INTERCEPTORS
} from '@angular/common/http';

import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { ModalModule } from 'ngx-bootstrap/modal';
import { ToasterModule } from 'angular2-toaster';

import {
  ToasterComponent,
  HttpProgressBarComponent
} from '@app/core/containers';
import { HttpProgressInterceptor } from '@app/core/services/http-progress.interceptor';

const MODULES: any[] = [
  HttpClientModule,
  TranslateModule.forRoot({
    loader: {
      provide: TranslateLoader,
      useFactory: HttpLoaderFactory,
      deps: [HttpClient]
    }
  }),
  ModalModule.forRoot(),
  ToasterModule.forRoot(),
  CommonModule
];

export const COMPONENTS = [ToasterComponent, HttpProgressBarComponent];

@NgModule({
  imports: MODULES,
  declarations: COMPONENTS,
  exports: COMPONENTS,
  providers: [
    {
      provide: HTTP_INTERCEPTORS,
      useClass: HttpProgressInterceptor,
      multi: true
    }
  ]
})
export class CoreModule {
  constructor(
    @Optional()
    @SkipSelf()
    parentModule: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        'CoreModule is already loaded. Import only once in main AppModule.'
      );
    }
  }
}

export function HttpLoaderFactory(http: HttpClient) {
  return new TranslateHttpLoader(
    http,
    `${environment.prefixAsset}/assets/i18n/`,
    '.json'
  );
}
