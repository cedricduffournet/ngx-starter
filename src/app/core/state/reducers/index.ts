import { InjectionToken } from '@angular/core';
import {
  createFeatureSelector,
  createSelector,
  ActionReducerMap,
  Action
} from '@ngrx/store';
import * as fromRouter from '@ngrx/router-store';
import * as fromLayout from '@app/core/state/reducers/layout.reducer';
import * as fromToaster from '@app/core/state/reducers/toaster.reducer';
import * as fromHttpProgress from '@app/core/state/reducers/http-progress.reducer';
import * as fromAuth from '@app/authentication/state/reducers';
import { RouterStateUrl } from '@app/core/utils/custom-route-serializer';

export interface State {
  layout: fromLayout.State;
  toaster: fromToaster.State;
  http: fromHttpProgress.State;
  router: fromRouter.RouterReducerState<RouterStateUrl>;
}

export const reducers = new InjectionToken<ActionReducerMap<State, Action>>(
  'Root reducers token',
  {
    factory: () => ({
      layout: fromLayout.reducer,
      toaster: fromToaster.reducer,
      http: fromHttpProgress.reducer,
      router: fromRouter.routerReducer
    })
  }
);

export const getRouterState = createFeatureSelector<
  fromRouter.RouterReducerState<RouterStateUrl>
>('router');

export const getPageTitle = createSelector(
  getRouterState,
  router => {
    if (router.state.data && router.state.data.title) {
      return router.state.data.title;
    }
    return '';
  }
);

export const getPreviousLink = createSelector(
  getRouterState,
  state => {
    if (state.state.data && state.state.data.previousLink) {
      return state.state.data.previousLink;
    }
    return null;
  }
);

export const getLayoutState = createFeatureSelector<State, fromLayout.State>(
  'layout'
);

export const getShowSidenav = createSelector(
  getLayoutState,
  fromLayout.getShowSidenav
);

export const getMenu = createSelector(
  getLayoutState,
  fromLayout.getMenu
);

export const getUserMenu = createSelector(
  getMenu,
  fromAuth.getLoggedUser,
  (menu, loggedUser) => {
    if (!loggedUser) {
      return [];
    }
    return menu.filter(item =>
      item.roles.some(role => loggedUser.roles.indexOf(role) !== -1)
    );
  }
);

export const getToasterState = createFeatureSelector<State, fromToaster.State>(
  'toaster'
);

export const getToasterParams = createSelector(
  getToasterState,
  fromToaster.getParams
);

export const getHttpState = createFeatureSelector<
  State,
  fromHttpProgress.State
>('http');

export const getHttpProcessList = createSelector(
  getHttpState,
  fromHttpProgress.getProcessList
);

export const isHttpProcessing = createSelector(
  getHttpProcessList,
  processList => processList.length !== 0
);
