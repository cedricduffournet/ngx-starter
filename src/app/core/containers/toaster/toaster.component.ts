import { Component, OnInit } from '@angular/core';
import { Store, select } from '@ngrx/store';

import { Observable } from 'rxjs';
import { skip } from 'rxjs/operators';
import {
  ToasterConfig,
  ToasterService
} from 'angular2-toaster/angular2-toaster';

import {
  defaultConfig,
  ToasterParams
} from '@app/core/containers/toaster/toaster.model';
import * as fromRoot from '@app/core/state/reducers';

@Component({
  selector: 'app-toaster',
  templateUrl: './toaster.component.html'
})
export class ToasterComponent implements OnInit {
  public toasterConfig: ToasterConfig = new ToasterConfig(defaultConfig);

  public toasterParams$: Observable<ToasterParams>;

  public constructor(
    private store: Store<fromRoot.State>,
    public toaster: ToasterService
  ) {}

  ngOnInit() {
    this.toasterParams$ = this.store.pipe(select(fromRoot.getToasterParams));

    this.toasterParams$.pipe(skip(1)).subscribe((params: ToasterParams) => {
      if (params.timeout && params.timeout > 0) {
        this.toasterConfig.timeout = params.timeout;
      }
      if (params.clear) {
        this.toaster.clear();
      }
      this.toaster.pop(params.type, params.title, params.body);
    });
  }
}
