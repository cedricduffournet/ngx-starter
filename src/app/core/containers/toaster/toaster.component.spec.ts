import { TestBed, ComponentFixture } from '@angular/core/testing';

import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { Store, MemoizedSelector } from '@ngrx/store';
import { ToasterModule } from 'angular2-toaster/angular2-toaster';

import { ToasterComponent, ToasterParams } from '@app/core/containers';
import * as fromToaster from '@app/core/state/reducers/toaster.reducer';
import * as fromRoot from '@app/core/state/reducers';

describe('ToasterComponent', () => {
  let fixture: ComponentFixture<ToasterComponent>;
  let component: ToasterComponent;
  let store: MockStore<fromRoot.State>;
  let params: MemoizedSelector<fromRoot.State, ToasterParams>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [ToasterModule.forRoot()],
      declarations: [ToasterComponent],
      providers: [provideMockStore()]
    });
    fixture = TestBed.createComponent(ToasterComponent);
    component = fixture.componentInstance;
    store = TestBed.get(Store);
    params = store.overrideSelector(
      fromRoot.getToasterParams,
      fromToaster.INITIAL_STATE.params
    );
  });

  it('should be created', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should call toaster pop', () => {
    spyOn(component.toaster, 'pop');
    fixture.detectChanges();
    params.setResult({
      title: 'test title',
      type: 'success',
      body: 'test body'
    });
    store.setState({} as any);
    expect(component.toaster.pop).toHaveBeenCalledWith(
      'success',
      'test title',
      'test body'
    );
  });

  it('should change timeout toaster config ', () => {
    fixture.detectChanges();
    params.setResult({
      title: 'test title',
      type: 'success',
      body: 'test body',
      timeout: 1000
    });
    store.setState({} as any);
    expect(component.toasterConfig.timeout).toBe(1000);
  });

  it('should clear toaster ', () => {
    spyOn(component.toaster, 'clear');
    fixture.detectChanges();
    params.setResult({
      title: 'test title',
      type: 'success',
      body: 'test body',
      clear: true
    });
    store.setState({} as any);
    expect(component.toaster.clear).toHaveBeenCalled();
  });
});
