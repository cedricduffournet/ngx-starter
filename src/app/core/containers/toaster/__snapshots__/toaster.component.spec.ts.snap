// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToasterComponent should be created 1`] = `
<app-toaster
  store={[Function MockStore]}
  toaster={[Function ToasterService]}
  toasterConfig={[Function ToasterConfig]}
  toasterParams$={[Function Store]}
>
  <toaster-container
    ng-reflect-toasterconfig="[object Object]"
  >
    <div
      class="toast-top-right"
      id="toast-container"
      ng-reflect-ng-class="toast-top-right"
    >
      
    </div>
  </toaster-container>
</app-toaster>
`;
